keps/0033-structured-parameters.md

```diff
@@ -0,0 +1,512 @@
+- Possible additional validation for our custom attributes e.g. parameter `trigger`
+
+The server-side of the proposal includes:
+- Updating `Operator` and `Instance` CRDs to use JSON-schema as the parameter format  
+- Providing a conversion-webhook to convert in-cluster running operators to the new format 
+   
+### Package Format
+
+JSON-schema is a [well-described](http://json-schema.org/understanding-json-schema/) format. However, here is a little example showing an existing `ZOOKEEPER_URI` and its new counterpart `Zookeeper.URI`:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+parameters:
+  - name: ZOOKEEPER_URI
+    displayName: "zookeeper cluster URI"
+    description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+    default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: true
+```
+
+```yaml
+type: object
+"$schema": "http://json-schema.org/draft-07/schema#"
+apiVersion: kudo.dev/v1beta2
+properties:
+  Zookeeper:
+    properties:
+      URI:
+        type: "string"
+        description: "host and port information for Zookeeper connection. e.g. zk:2181,zk2:2181,zk3:2181"
+        default: "zookeeper-instance-zookeeper-0.zookeeper-instance-hs:2181"
+    required: ["URI"]
+```
+
+KUDO will support both package formats for the future. At some point the old format might become deprecated, but this is not planned yet.
+
+Note: The apiVersions here (`v1beta1` and `v2beta2`) correspond to the CRD versions, but are separate definitions, even if they seem to be similar in certain aspects.
+
+### CRDs
+
+Normally, per Kubernetes conventions, CRDs would evolve over released versions without a breaking change from one version to the next, allowing clients to gradually convert to new CRD versions. This requires for a CRD to be convertible from the old version to new _and vice versa_ for all clients (new and old) to be able to read/write both old and new CRD versions. However, such a conversion in both directions will not be possible with this change, as JSON-schema supports a lot of configurations which will not be mappable from a flat list. 
+
+There will be a hard break for clients and users of CRDs from `v1beta1` to the new `v1beta2`. We will provide a migration path for existing installed CRDs, but the clients accessing these CRDs will need to be updated.
+
+#### Conversion
+
+To migrate the existing CRDs, we will implement a Conversion Webhook that only allows `v1beta1` to `v1beta2` conversion, not the other way around. When KUDO is upgraded, it will:
+
+1. Deploy the new CRD version where the new (`v1beta2`) version will be the _only_ served and stored version:
+    - `v1beta1` will be still in the list, but with `storage=false, served=false`
+    - `v1beta2` will be marked as `storage=true, served=true`
+2. Deploy the new manager with the conversion webhook
```

Comment:
I would personally deploy manager before the crd? so there's already the conversion implementation in place when the crd comes.

Reply:
Hmmm. I'm not sure. Starting the manager first would mean the manager would try to start reconciliations, or maybe couldn't even start because it tries to register a watch on Instance with `v1beta1` which is unknown to the cluster.

Maybe this is the reason that projects separate the manager and the conversion webhook? That's another option. I would prefer to keep it in the manager, but maybe we have to make it a separate deployment?

Reply:
hmm I see, maybe. but that would make our upgrade process even more complicated. Can we even right now express the steps like this easily?

Reply:
Yeah, I think the steps here can be implemented pretty easy. I've updated the section a little bit.

Reply:
> I would personally deploy manager before the crd? 

Like Andreas mentioned, you need the new CRD version before the manager can register itself to watch them, right? And the resources has to be converted too at this moment. So the whole update process will be something like:
1. Kill the manager (or maybe step 3 can do it for the user?) 
2. `kubectl krew upgrade kudoctl` to upgrade the kudoctl
3. `kudo init --update` to install new CRDs, run the new manager+webhook, and trigger the migration

Reply:
`kudo init --upgrade` kills the manager already, so there's no need to manually shut it down. The whole proces described here is internal to the `kudo init --upgrade`